### ISSUE TITLE
GenFacades don't support rewriting Portable PDBs

### ISSUE THREAD

**Issue by joperezr** [dotnet/buildtools#705]
Currently our x-plat builds for corefx generate portable PDBs for each library. After generating each one, we run GenFacades on top of the binary and symbols to add some type-forwards. Even though this works fine in Windows(because Windows doesn't generate Portable PDBs), it doesn't work on x-plat because CCI (GenFacades dependency) doesn't support Portable PDB rewriting. This issue is to track that once CCI adds support for them, we should make sure that GenFacades work as expected and that the symbols can now be used for debugging.

CC: @weshaggard 
related: dotnet/corefx#5733 dotnet/corefx#5825

**Comment by nguerrera** [comment]
I suggest changing GenFacades to a source generator that generates type-forwards in intermediate directory, then just csc together with manual source (if any) to emit final dll and pdb. That is probably less work than portable PDB support in CCI (who's signing up to do that?) and also future proof against further binary rewriting fragility. cc @tmat

**Comment by tmat** [comment]
Also it will make the build deterministic without the need to make CCI deterministic.

**Comment by mjsabby** [comment]
Is someone working on CCI getting Portable PDB rewriting capabilities?

**Comment by tmat** [comment]
@mjsabby Not that I know of.

**Comment by dleeapho** [comment]
Triage: Genfacades is blocking signing in the standard repo. We might want to consider moving genfacades to a Roslyn rewriter.

**Comment by tmat** [comment]
@dleeapho Yes, I believe that'd be the right direction. Any tool that can generate source code that gets fed into the standard compiler task is better of doing so than IL/metadata rewriting.

**Comment by alexperovich** [comment]
As an FYI: A source generator will not work with specific type forwards that GenFacades needs to make. Specifically nested classes like `List<T>.Enumerator`. `[TypeForwardedTo(typeof(List<T>.Enumerator))]` doesn't work, and `[TypeForwardedTo(typeof(List<T>))]` doesn't also forward the enumerator.

**Comment by nguerrera** [comment]
Yes it does

**Comment by nguerrera** [comment]
Forwarding outer type with attribute auto forwards nested types

**Comment by alexperovich** [comment]
Was that added recently? I didn't see the nested types forwarded when I tried it.

**Comment by nguerrera** [comment]
No. It's been there since pre-roslyn.


``` C#
using System.Runtime.CompilerServices;
using System.Collections.Generic;
[assembly:TypeForwardedTo(typeof(List<>))]
```

```
.class extern forwarder System.Collections.Generic.List`1
{
  .assembly extern mscorlib
}
.class extern SynchronizedList 
{
  .class extern System.Collections.Generic.List`1
}
.class extern Enumerator
{
  .class extern System.Collections.Generic.List`1
}
```

NOTE: What looks like a "nested forwarder" in CCI API is an `ExportedType` with `Implementation` set to the enclosing `ExportedType` in metadata. The `IsTypeForwarder` flag is not actually set, hence no `forwarder`keyword in IL, but the metadata above is exactly what GenFacades needs to emit.

**Comment by weshaggard** [comment]
Yes this has always worked and it should source generator will work for this scenario we just need to write the tool and figure out how to best plug it into our build pipeline.

**Comment by danmosemsft** [comment]
@joperezr shall we close in favor of arcade? https://github.com/dotnet/arcade/issues/1584
